sdk/attestation/azure-security-attestation/src/test/java/com/azure/security/attestation/AttestationClientTestBase.java

```diff
@@ -0,0 +1,317 @@
+        } catch (NoSuchAlgorithmException e) {
+            throw logger.logThrowableAsError(new RuntimeException(e));
+        }
+
+        PrivateKey privateKey;
+        try {
+            privateKey = keyFactory.generatePrivate(keySpec);
+        } catch (InvalidKeySpecException e) {
+            throw logger.logExceptionAsError(new RuntimeException(e));
+        }
+
+        return new RSASSASigner(privateKey);
+    }
+
+    /**
+     * Find the signing certificate associated with the specified SignedJWT token.
+     *
+     * This method depends on the token
+     * @param token - MAA generated token on which to find the certificate.
+     * @param client - Http Client used to retrieve signing certificates.
+     * @param clientUri - Base URI for the attestation client.
+     * @return X509Certificate which will have been used to sign the token.
+     */
+    Mono<X509Certificate> getSigningCertificateByKeyId(SignedJWT token, HttpClient client, String clientUri) {
+        AttestationClientBuilder builder = getBuilder(client, clientUri);
+        return builder.buildSigningCertificatesAsyncClient().get()
+            .handle((keySet, sink) -> {
+                final CertificateFactory cf;
+                try {
+                    cf = CertificateFactory.getInstance("X.509");
+                } catch (CertificateException e) {
+                    sink.error(logger.logThrowableAsError(e));
+                    return;
+                }
+
+                final String keyId = token.getHeader().getKeyID();
+                boolean foundKey = false;
+                for (JsonWebKey key : keySet.getKeys()) {
+                    if (keyId.equals(key.getKid())) {
+                        final Certificate cert;
+                        try {
+                            cert = cf.generateCertificate(base64ToStream(key.getX5C().get(0)));
+                            foundKey = true;
+                        } catch (CertificateException e) {
+                            sink.error(logger.logThrowableAsError(e));
+                            return;
+                        }
+
+                        assertTrue(cert instanceof X509Certificate);
+                        sink.next((X509Certificate) cert);
+                    }
+                }
+                if (!foundKey) {
+                    sink.error(logger.logThrowableAsError(new RuntimeException(String.format(
+                        "Key %s not found in JSON Web Key Set", keyId))));
+                }
+            });
+    }
+
+    String getIsolatedSigningCertificate() {
+        String signingCertificate = Configuration.getGlobalConfiguration().get("isolatedSigningCertificate");
+        if (signingCertificate == null) {
+            // Use a pre-canned signing certificate captured at provisioning time.
+            signingCertificate = "MIIC+DCCAeCgAwIBAgIITwYg6gewUZswDQYJKoZIhvcNAQELBQAwMzExMC8GA1UEAxMoQXR0ZXN0YXRpb25Jc"
```

Comment:
Is this okay to put certificate in this file from security point of view ?

Reply:
The certificate has to go somewhere. It's either going to be in the recorded session or it's going to be in the source code, but the service requires that the client present a valid X.509 certificate.

So we either don't bother testing most of the service functionality in playback tests or we add the certificate (and it's associated key).

In this case, it's a self signed certificate wrapped around a generate RSA key pair (that's the key returned by the `getIsolatedSigningKey`, which was generated for this purpose.

It's just a certificate and a key - they don't protect actual resources (I guess technically they are the isolated signing key for an attestation service instance I used for testing when I created these tests, but I'm not sure that even with that caveat, they have any practical value - I delete that instance regularly.